### ISSUE TITLE
Hybrid offset?

### ISSUE THREAD

**Issue by t47io** [brigade/react-waypoint#167]
I was reading the source code of `computeOffsetPixels.js` and now understand how it works. I would like to raise an discussion on a possible expansion for the **"hybrid offset"**, e.g. allowing `100% - 50px`:

I found one use case in my project where I use `react-waypoint` as a scrollspy agent. I have a fixed-height `<nav>` at the top of the page. And I want to trigger something when the user scroll down to a point where a `<div>` is right 'at  flush' to the bottom of the `<nav>`.

Since I want the above behavior when scrolling down, I need to use `bottomOffset`, not `topOffset` (right?). So basically I need `100% - 50px`-ish value for `bottomOffset`, which is not supported. Such functionality will be handy, since `_handleScroll()` gets the bound from `_getBounds()` fresh, and that takes care of resizing issues.

**Comment by jmeas** [comment]
Hi @t47io !

You're right that we don't currently support a hybrid offset, but you're in luck, because I think the API that we have supports your use case. In fact, by the sound of it, that's the exact use case I needed to support when I helped implement the current offset API.

> Since I want the above behavior when scrolling down, I need to use bottomOffset, not topOffset (right?).

Nope, it's still `topOffset`. You can think of `topOffset` as controlling the location of the boundary at the top of the page. `bottomOffset` controls the boundary near the bottom of the page. In this case, you're shifting the location of that top boundary, so, accordingly, `topOffset` is the best bet. If you had a fixed footer, for example, then you'd want to use `bottomOffset`.

To know what value we need to set for `topOffset`, we can reference [the documentation](https://github.com/brigade/react-waypoint#offsets-and-boundaries):

> The `topOffset` and `bottomOffset` properties can adjust the placement of these boundaries. By default, the offset is '0px'. If you specify a positive value, then the boundaries will be pushed inward, toward the center of the page. If you specify a negative value for an offset, then the boundary will be pushed outward from the center of the page. 

Based on this, it sounds like what you need is a positive pixel value of `topOffset` equal to the height of your nav bar. If your nav bar is 50px, then `topOffset="50px"` should do the trick.

Also, for some historical context, we did consider a hybrid API, similar to what you suggested.  You can read more about that conversation here: https://github.com/brigade/react-waypoint/issues/77#issuecomment-223705419 

Let me know if that helps!

**Comment by t47io** [comment]
Hi @jmeas Thanks a lot for the fast response. I think your suggestion works for a `<Waypoint/>` without children, in which I set `topOffset` to 50px, and trigger it with `onLeave` (that's when it hits the top of viewport). But my use case involves children. And quote:

> The `onEnter` callback will be called when _any_ part of the child is visible in the viewport. The `onLeave` callback will be called when _all_ of the child has exited the viewport.

Let me draw a diagram to clarify

**Comment by t47io** [comment]
![screen shot 2017-03-26 at 03 43 31 pm](https://cloud.githubusercontent.com/assets/4424215/24335921/fdb1e300-123a-11e7-80a3-d0332048c7bf.png)

So basically, how could I detect the following 4 events separately:
* [TOP of <div>] hits [BOTTOM of viewport]
* [TOP of <div>] hits [TOP of viewport]
* [BOTTOM of <div>] hits [BOTTOM of viewport]
* [BOTTOM of <div>] hits [BOTTOM of viewport]

**Comment by jmeas** [comment]
@t47io , as I understand the current API, the callbacks you're looking for is not provided by Waypoint right now, when there are children. To get that information, you would need to use two separate Waypoints, one for the top, and one for the bottom.

The reason is that a Waypoint, whether it has children or not, does not have the concept of a top nor a bottom. There's just a concept of whether the waypoint is in view or not. This no doubt stems from the history of a Waypoint without children, in which the "top" and "bottom" are the same thing (because the waypoint is just a line).

As an immediate solution, I'd recommend using two waypoints rather than the children. If you're not satisfied with that solution, I'd be curious to hear your proposal for an API that would introduce this new concept of a waypoint having a top and a bottom.

**Comment by t47io** [comment]
@jmeas Yes, sandwich should work. And this is what I came up with:
![screen shot 2017-03-26 at 04 23 00 pm](https://cloud.githubusercontent.com/assets/4424215/24336256/8f7c68b4-1240-11e7-9ac7-88bf735b99d8.png)

The only issue I thought of is the following scenario where:
* Load the page, scroll to the inside the `<div/>` (very long div)
* Refresh the page, and now we "land in" the middle of it, without scrolling, thus without `onEnter` or `onLeave`

So we need `onPositionChange`. However, I found that when the page loads, `onPositionChange` reports `previousPosition=undefined` (only once), thus the scrollspy need an "initial scrolling" in order to capture the current position. Any thoughts?

**Comment by t47io** [comment]
oh wait, wrong diagram

**Comment by jmeas** [comment]
> So we need onPositionChange. However, I found that when the page loads, onPositionChange reports previousPosition=undefined (only once), thus the scrollspy need an "initial scrolling" in order to capture the current position. Any thoughts?

@t47io , this makes sense to me. This could serve as further motivation for us to introduce some sort of `onMount` callback, sending the position on page load. A thread discussing that is here: #164 . However, in #155 , which also discusses this problem, a workaround for that problem is described. I can't look into it just this moment to see if that workaround would work for you, but if you had the time it could be worthwhile.

**Comment by t47io** [comment]
@jmeas The previous one still required "hybrid". This should work: 

![screen shot 2017-03-26 at 04 43 11 pm](https://cloud.githubusercontent.com/assets/4424215/24336394/52b7662e-1243-11e7-8c99-b2bc129ffbae.png)

For the `onMount` problem, I found that although `previous=undefined`, we still have `current=inside`, So the `<div/>` body can be wrapped by a third `<Waypoint>` and only responsible for listening to the `onMount` position. Something like:
```jsx
const MyComponent = ({ children }) => (
  <div>
    <Waypoint /> /* do the id="1" job */
    <Waypoint
      onPositionChange={({ currentPosition, previousPosition }) => {
        if (typeof previousPosition === 'undefined' && currentPosition === Waypoint.inside) {
          // do stuff
        }
      }}
    >
      {children}
    </Waypoint>
    <Waypoint /> /* do the id="2" job */
  </div>
);
```

**Comment by t47io** [comment]
Is there an interest for including this as a `<Scrollspy/>` in `react-waypoint`? I can play with a PR

**Comment by jmeas** [comment]
That's awesome that you found a workaround! I'm glad our API supports that use case. However, I'm not too satisfied with it. It seems a little messier than I'd like.

I'd much prefer to see one of:

- a cleaner solution using the current API 
- the introduction of a concept of a top/bottom of a waypoint, so that this works with `children`
- the introduction of an `onMount` (or whatever) callback, so that this works with two waypoints wrapping the content

> Is there an interest for including this as a <Scrollspy/> in react-waypoint? I can play with a PR

My preference would be to fix the issue with one of the above solutions over something like the Scrollspy code you wrote. If it works, that's awesome, but I think we can do better (even if it requires updating the API of this lib).

@trotzig / @lencioni , this thread is a lot to take in, but I'd be curious to hear your thoughts whenever you've got time to review this.

**Comment by t47io** [comment]
Yes, agreed. This is only a workaround, not any elegant solution. I can imagine my proposed component like:
```
topOffset: shapeOf({
  viewTop: oneOfType([number, string]),
  viewBottom: oneOfType([number, string]),
}),
bottomOffset: shapeOf({
  viewTop: oneOfType([number, string]),
  viewBottom: oneOfType([number, string]),
}),
onTopEnter: func,
onTopLeave: func,
onBottomEnter: func,
onBottomLeave: func,
scrollableAncestor: any,
fireOnRapidScroll: bool,
debug: bool,
```
> top/bottom meaning of the `<div>` which is intuitive, and viewTop/Bottom for viewport.

Handlers are triggered by the internal `<Waypoint/>` # 1,2,3, with same arguments. The caveat is, indeed, it does not address the "awareness of top/bottom" problem. Also, it would be ideal to have a "real-time" `onPositionChange`, sort of like the one here https://github.com/gilbox/react-spark-scroll#callback-on-scroll-event
